src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs

```diff
@@ -0,0 +1,414 @@
+                        {
+                            Debug.Assert(_first != null);
+                            _first[i] = values[index + (int)AggregationType.First];
+                        }
+                        goto case AggregationType.Last;
+                    case AggregationType.Percentile:
+                        throw new InvalidOperationException($"Cannot use aggregation method '{nameof(AggregationType.Percentile)}' on rolled-up time series");
+                    default:
+                        throw new ArgumentOutOfRangeException("Unknown aggregation operation: " + Aggregation);
+                }
+
+                _count[i] += (long)values[index + (int)AggregationType.Count];
+            }
+
+            TotalCount += (long)values[(int)AggregationType.Count];
+        }
+
+        public IEnumerable<double> GetFinalValues(DateTime? from, DateTime? to, double? scale = null)
+        {
+            if (_finalValues != null)
+                return _finalValues;
+
+            switch (Aggregation)
+            {
+                case AggregationType.Min:
+                case AggregationType.Max:
+                case AggregationType.First:
+                case AggregationType.Last:
+                case AggregationType.Sum:
+                    break;
+                case AggregationType.Count:
+                    return _count;
+                case AggregationType.Average:
+                    for (int i = 0; i < _values.Count; i++)
+                    {
+                        if (_count[i] == 0)
+                        {
+                            _values[i] = double.NaN;
+                            continue;
+                        }
+
+                        _values[i] = _values[i] / _count[i];
+                    }
+                    break;
+                case AggregationType.Percentile:
+                    return _finalValues = GetPercentile(scale ?? 1);
+                case AggregationType.Slope:
+                    Debug.Assert(_values.Count == _first.Count, "Invalid aggregation data");
+                    Debug.Assert(from.HasValue && to.HasValue, "Missing from/to");
+
+                    var deltaX = (to.Value.Ticks - from.Value.Ticks) / 10_000; // distance in milliseconds
+                    for (int i = 0; i < _values.Count; i++)
+                    {
+                        var deltaY = _values[i] - _first[i]; // last - first
+                        _values[i] = deltaY / deltaX;
+                    }
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException("Unknown aggregation operation: " + Aggregation);
+            }
+
+            if (scale.HasValue == false) 
+                return _finalValues = _values;
+
+            return _finalValues = _values.Select(x => x * scale.Value);
+        }
+
+        private IEnumerable<double> GetPercentile(double scale)
+        {
+            foreach (var list in _rankedValues)
```

Comment:
This rounds down, right? Do we need some manner to adjust the value based on the distance?
So if we have [1,2,4] and we ask for 75% percentile, we'll get [2] here. But that actually represent the 50%.
Should we adjust that so it would be 3 as the result?

Reply:
No actually we round up : `index = (int)Math.Ceiling(_percentileFactor * list.Count)`
so in your example we'll get `4` as the result. (`2` is wrong by any defenition)

There are different defenitions for percentile and methods for calculting percentile.
By the defenition that I chose to implement the result is always a part of the data set.
So if we want to ajust the result to be `3` we need to use a different method..

I wrote more about it in the issue page : https://issues.hibernatingrhinos.com/issue/RavenDB-15020#focus=Comments-67-51024.0-0

Reply:
Okay, that make sense.
If it isn't hard, let's go with extrapolation. Note that we need to document this.

Reply:
done